engine/glfw/lib/android/android_joystick.c

```diff
@@ -1,37 +1,459 @@
 // Copyright 2020 The Defold Foundation
 // Licensed under the Defold License version 1.0 (the "License"); you may not use
 // this file except in compliance with the License.
-// 
+//
 // You may obtain a copy of the License, together with FAQs at
 // https://www.defold.com/license
-// 
+//
 // Unless required by applicable law or agreed to in writing, software distributed
 // under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 // CONDITIONS OF ANY KIND, either express or implied. See the License for the
 // specific language governing permissions and limitations under the License.
 
 #include "internal.h"
+#include "android_joystick.h"
+#include "android_log.h"
+
+#include <jni.h>
+
+extern struct android_app* g_AndroidApp;
+
+static int glfwAndroidJoystickPresent( int joy )
+{
+    if (joy >= 0 && joy <= GLFW_JOYSTICK_LAST && _glfwJoy[joy].State == GLFW_ANDROID_GAMEPAD_CONNECTED)
+    {
+        return GL_TRUE;
+    }
+    return GL_FALSE;
+}
+
+int glfwAndroidFindJoystick(const int32_t deviceId)
+{
+    int32_t joystickIndex;
+    for (joystickIndex = 0; joystickIndex <= GLFW_JOYSTICK_LAST; joystickIndex++)
+    {
+        if (_glfwJoy[joystickIndex].DeviceId == deviceId)
+        {
+            return joystickIndex;
+        }
+    }
+    return -1;
+}
+
+static int32_t glfwAndroidConnectJoystick(int32_t deviceId, char* deviceName)
+{
+    int n;
+    int32_t joystickIndex = -1;
+    for (joystickIndex = 0; joystickIndex <= GLFW_JOYSTICK_LAST; joystickIndex++)
+    {
+        if (_glfwJoy[joystickIndex].State == GLFW_ANDROID_GAMEPAD_DISCONNECTED)
+        {
+            _glfwJoy[joystickIndex].State = GLFW_ANDROID_GAMEPAD_CONNECTED;
+            _glfwJoy[joystickIndex].DeviceId = deviceId;
+            _glfwJoy[joystickIndex].NumAxes = GLFW_ANDROID_GAMEPAD_NUMAXIS;
+            _glfwJoy[joystickIndex].NumButtons = GLFW_ANDROID_GAMEPAD_NUMBUTTONS;
+            _glfwJoy[joystickIndex].Axis = (float *)malloc(sizeof(float) * _glfwJoy[joystickIndex].NumAxes);
+            _glfwJoy[joystickIndex].Button = (unsigned char *)malloc(sizeof(char) * _glfwJoy[joystickIndex].NumButtons );
+            strncpy(&_glfwJoy[joystickIndex].DeviceName, deviceName, DEVICE_NAME_LENGTH);
+
+            // Clear joystick state
+            for(n = 0; n < _glfwJoy[joystickIndex].NumAxes; ++n)
+            {
+                _glfwJoy[joystickIndex].Axis[n] = 0.0f;
+            }
+            for(n = 0; n < _glfwJoy[joystickIndex].NumButtons; ++n)
+            {
+                _glfwJoy[joystickIndex].Button[n] = GLFW_RELEASE;
```

Comment:
... which might or might not work well with such constants, but usually they're chosen to be default==false==0

Reply:
Code gets slightly shorter, but might still be nice since we want smaller executables, and the readability isn't really affected imho.

https://godbolt.org/z/3nxMY79j5

Reply:
Good point. GLFW_RELEASE is 0 so memsetting to 0 works.